java-checks/src/main/java/org/sonar/java/checks/TypeUpperBoundNotFinalCheck.java

```diff
@@ -0,0 +1,105 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeParameterTree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.WildcardTree;
+
+@Rule(key = "S4968")
+public class TypeUpperBoundNotFinalCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.TYPE_PARAMETER, Tree.Kind.EXTENDS_WILDCARD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.TYPE_PARAMETER)) {
+      handleBounds(((TypeParameterTree) tree).bounds(), tree);
+    } else if (tree.is(Tree.Kind.EXTENDS_WILDCARD)) {
+      handleBounds(Collections.singletonList(((WildcardTree) tree).bound()), tree);
+    }
+  }
+
+  private void handleBounds(List<TypeTree> bounds, Tree treeToReport) {
+    for (TypeTree bound : bounds) {
+      if (reportIssueIfBoundIsFinal(bound, treeToReport))
+        return;
+    }
+  }
+
+  private boolean reportIssueIfBoundIsFinal(TypeTree bound, Tree treeToReport) {
+    if (bound.is(Tree.Kind.IDENTIFIER)) {
+      if (isFinal((IdentifierTree) bound) && !inOverridingMethodDeclaration(bound)) {
+        reportIssue(treeToReport, "Replace this type parametrization by the 'final' type.");
+        return true;
+      }
+    } else if (bound.is(Tree.Kind.PARAMETERIZED_TYPE)) {
+      ParameterizedTypeTree type = (ParameterizedTypeTree) bound;
+      if (reportIssueIfBoundIsFinal(type.type(), treeToReport)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static boolean isFinal(IdentifierTree bound) {
+    return bound.symbol().isFinal();
+  }
+
+  // Returns true if 'type' is part of the signature of an overriding method.
+  private static boolean inOverridingMethodDeclaration(TypeTree type) {
+    Optional<MethodTree> method = getMethodDeclaration(type);
+    if (method.isPresent()) {
+      return isOverriding(method.get());
+    } else {
+      return false;
+    }
+  }
+
+  // Returns the method where 'type' is in its signature, if any.
+  private static Optional<MethodTree> getMethodDeclaration(TypeTree type) {
```

Comment:
We may start from "treeToReport" instead of the type, to spare a few parent() calls.